src/analysis/multiSpecies/microbiomeModelingToolbox/additionalAnalysis/analyseObjectiveShadowPrices.m

```diff
@@ -0,0 +1,191 @@
+            model=changeObjective(model,objectiveList{j,1});
+            FBAsolution=solveCobraLP(model);
+            solutionsTemp{j,1}=FBAsolution;
+            end
+        end
+        solutions(:,i)=solutionsTemp;
+    end
+else
+    % without parallelization
+    for i=1:size(modelList,1)
+        ShadowPrices{1,i+2}=modelIDs{i,1};
+        model=modelList{i,1};
+        % Compute the solutions for all objectives
+        for j=1:size(objectiveList,1)
+            % optimize for the objective if it is present in the model
+            if ~isempty(find(ismember(model.rxns,objectiveList{j,1})))
+                model=changeObjective(model,objectiveList{j,1});
+                FBAsolution=solveCobraLP(model);
+                solutions{j,i}=FBAsolution;
+            end
+        end
+    end
+end
+% Extract all shadow prices and save them in a table
+for i=1:size(modelList,1)
+    model=modelList{i,1};
+    for j=1:size(objectiveList,1)
+        % get the computed solutions
+        if ~isempty(solutions{j,i})
+            FBAsolution=solutions{j,i};
+            % verify that a feasible solution was obtained
+            if FBAsolution.stat==1
+                [extractedShadowPrices]=extractShadowPrices(model,FBAsolution,SPDef);
+                for k=1:size(extractedShadowPrices,1)
+                    % check if the metabolite relevant for this objective
+                    % function is already in the table
+                    findMet=find(strcmp(ShadowPrices(:,1),extractedShadowPrices{k,1}));
+                    findObj=find(strcmp(ShadowPrices(:,2),objectiveList{j,1}));
+                    if ~isempty(intersect(findMet,findObj))
+                        % Add the shadow price for this model
+                        ShadowPrices{intersect(findMet,findObj),i+2}=extractedShadowPrices{k,2};
+                    else
+                        % Add a new row for this metabolite and objective function with the shadow price for this model
+                        newRow=size(ShadowPrices,1)+1;
+                        ShadowPrices{newRow,1}=extractedShadowPrices{k,1};
+                        ShadowPrices{newRow,2}=objectiveList{j,1};
+                        ShadowPrices{newRow,i+2}=extractedShadowPrices{k,2};
+                    end
+                end
+            end
+        end
+    end
+end
+end
+
+function [extractedShadowPrices]=extractShadowPrices(model,FBAsolution,SPDef)
+% Finds all shadow prices in a solution computed for a COBRA model
+% structure that indicate the metabolite is relevant for the flux through the objective function.
+
+extractedShadowPrices={};
+% Find all shadow prices (negative or positive depending on variable
+% SPDef)
+cnt=1;
+for i=1:length(model.mets)
+    % Do not include slack variables
+    if ~strncmp('slack_',model.mets{i},6)
+        if strcmp(SPDef,'Negative')
+            if FBAsolution.dual(i)  <0 && abs(FBAsolution.dual(i))>0.00000001
```

Comment:
May you use here a `tol` variable and set it with scientific notation?

Reply:
@laurentheirendt I am not sure what this means either.